### PR TITLE
Plane: send attitude target message implementation (for quadplanes)

### PR DESCRIPTION
**Pull request by miodine**
Minor mod to the communication routine, whose purpose is to expose the target attitude for a plane operating in VTOL mode.

Might be useful for external in-flight PID tuning applications.

### CODE REVIEW

File: ArduPlane/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -157,6 +157,48 @@ void GCS_MAVLINK_Plane::send_attitude() const
...
+    float target_y = 0.0f;
+    float throttle = 0.0f;
+    Quaternion quat(0.0f,0.0f,0.0f,0.0f);
+
+#if HAL_QUADPLANE_ENABLED
+    plane.quadplane.attitude_control->get_throttle_in();
+
+    if (!(plane.g2.flight_options & FlightOptions::GCS_REMOVE_TRIM_PITCH_CD)) {
+        target_p -= radians(plane.g.pitch_trim_cd * 0.01f);
+    }
```
Comment (IamPete1) [inline ArduPlane/GCS_Mavlink.cpp:173]: This does not do anything.

File: ArduPlane/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -157,6 +157,48 @@ void GCS_MAVLINK_Plane::send_attitude() const
...
+void GCS_MAVLINK_Plane::send_attitude_target() {
+
+    float target_r = 0.0f;
+    float target_p = 0.0f;
+    float target_y = 0.0f;
+    float throttle = 0.0f;
+    Quaternion quat(0.0f,0.0f,0.0f,0.0f);
+
+#if HAL_QUADPLANE_ENABLED
+    plane.quadplane.attitude_control->get_throttle_in();
```
Comment (IamPete1) [inline ArduPlane/GCS_Mavlink.cpp:169]:
```suggestion
    throttle = plane.quadplane.attitude_control->get_throttle_in();
```

**Review by IamPete1 — DISMISSED** [review]
This seems to have collected some extra unrelated changes. It only need to be sent if quadplane attitude control is active, 


You should be able to more or less copy exactly the copter implementation.

File: ArduPlane/GCS_Mavlink.cpp (hunk 1 above)
Comment (miodine) [reply ArduPlane/GCS_Mavlink.cpp]: right, deleted

**Review by IamPete1 — COMMENTED** [review]
You can use `last_att_control_ms` to check if its valid to report, like we do for PID logging:

https://github.com/ArduPilot/ardupilot/blob/fcd4152b30aa1f50b110303e7cae1f9f35f5e645/ArduPlane/Log.cpp#L30

So if not active you can early return

Then you can just copy and paste the copter implementation to get the values, just swapping `copter.` for `plane.quadplane.`

https://github.com/ArduPilot/ardupilot/blob/fcd4152b30aa1f50b110303e7cae1f9f35f5e645/ArduCopter/GCS_Mavlink.cpp#L98-L100

File: ArduPlane/GCS_Mavlink.cpp [hunk 3]
```diff
@@ -159,33 +159,25 @@ void GCS_MAVLINK_Plane::send_attitude() const
 
 void GCS_MAVLINK_Plane::send_attitude_target() {
 
-    float target_r = 0.0f;
-    float target_p = 0.0f;
-    float target_y = 0.0f;
-    float throttle = 0.0f;
-    Quaternion quat(0.0f,0.0f,0.0f,0.0f);
+    // Check if the attitude target is valid for reporting
+    if (AP_HAL::millis() - plane.quadplane.last_att_control_ms > 100) return;
```
Comment (IamPete1) [inline ArduPlane/GCS_Mavlink.cpp:163]: You can stash millis in a const and use it down on 175 to save getting it twice. Return should be on the next line per the style guide.
```suggestion
    const uint32_t now = AP_HAL::millis()
    if (now  - plane.quadplane.last_att_control_ms > 100) {
        return;
    }
```

File: ArduPlane/GCS_Mavlink.cpp [hunk 4]
```diff
@@ -159,33 +159,25 @@ void GCS_MAVLINK_Plane::send_attitude() const
...
-        throttle = plane.quadplane.attitude_control->get_throttle_in();
-        quat = plane.quadplane.attitude_control->get_attitude_target_quat();
-    }
-
-    const float quat_out[4] = {quat.q1, quat.q2, quat.q3, quat.q4};
     const uint16_t typemask = 0; 
 
     mavlink_msg_attitude_target_send(
         chan,
         AP_HAL::millis(),       // time since boot (ms)
```
Comment (IamPete1) [inline ArduPlane/GCS_Mavlink.cpp:175]:
```suggestion
        now,       // time since boot (ms)
```

File: ArduPlane/GCS_Mavlink.cpp (hunk 3 above)
Comment (miodine) [reply ArduPlane/GCS_Mavlink.cpp:163]: Right, of course, I completely missed the other function call

**Review by IamPete1 — COMMENTED** [review]
Code changes now look good, just need to drop the logging changes and squash the commits into one with a sensible name. `Plane: send attitude target for quadplanes` or similar.

File: ArduPlane/GCS_Mavlink.cpp [hunk 5]
```diff
@@ -157,6 +157,35 @@ void GCS_MAVLINK_Plane::send_attitude() const
         omega.z);
 }
 
+void GCS_MAVLINK_Plane::send_attitude_target() {
+
```
Comment (IamPete1) [inline ArduPlane/GCS_Mavlink.cpp:161]: Sorry, I just remembered something. We need to define this out if quadplane is not enabled.
```suggestion
#if HAL_QUADPLANE_ENABLED
```

and a `#endif // HAL_QUADPLANE_ENABLED` at the end

File: ArduPlane/GCS_Mavlink.cpp [hunk 6]
```diff
@@ -157,6 +157,35 @@ void GCS_MAVLINK_Plane::send_attitude() const
...
+    // Check if the attitude target is valid for reporting
+    const uint32_t now = AP_HAL::millis()
```
Comment (IamPete1) [inline ArduPlane/GCS_Mavlink.cpp:163]:
```suggestion
    const uint32_t now = AP_HAL::millis();
```

**Comment by miodine** [comment]
I did as it was requested - I applied the most recent suggestions, reverted any chages done in the logging part, and squashed the commits altogether.

**Review by IamPete1 — APPROVED** [review]
I think this looks good. Its not added to any default stream rate, so it will not be sent by default, but it can be requested.

**Comment by IamPete1** [comment]
@miodine I have marked for the dev cal which is this evening on Discord, your welcome to come along. 

https://ardupilot.org/dev/docs/ardupilot-discord-server.html#weekly-dev-calls

**Comment by miodine** [comment]
All right, then I'll await the results of your discussion. Sorry for my clumsiness with all that, and thanks for the guidance

File: ArduPlane/GCS_Mavlink.cpp [hunk 7]
```diff
@@ -157,6 +157,39 @@ void GCS_MAVLINK_Plane::send_attitude() const
...
+void GCS_MAVLINK_Plane::send_attitude_target() 
+{
+#if HAL_QUADPLANE_ENABLED
+    // Check if the attitude target is valid for reporting
+    const uint32_t now = AP_HAL::millis();
+    if (now  - plane.quadplane.last_att_control_ms > 100) {
+        return;
+    }
+    
+    const Quaternion quat  = plane.quadplane.attitude_control->get_attitude_target_quat();
```
Comment (tridge) [inline ArduPlane/GCS_Mavlink.cpp:169]: should be a const ref in AC_AttitudeControl.h, and ref here

File: ArduPlane/GCS_Mavlink.cpp [hunk 8]
```diff
@@ -157,6 +157,39 @@ void GCS_MAVLINK_Plane::send_attitude() const
...
+    const Vector3f ang_vel = plane.quadplane.attitude_control->get_attitude_target_ang_vel();
```
Comment (tridge) [inline ArduPlane/GCS_Mavlink.cpp:170]: should be const ref

**Review by tridge — CHANGES_REQUESTED** [review]
fix refs in copter too?

**Comment by tridge** [comment]
this needs to be against master, not plane-4.4

**Comment by tridge** [comment]
@IamPete1 will rework this as a PR against master

**Pull request closed without merging**